Pitch Bending Directly in the Piano Roll

regulus79 opened the pull request:
This PR modifies how the Detuning tool works in the piano roll to let the user modify the automation curve right on the notes. This is actually fairly simple, as automation clips already contain the functions for dragging and removing points like in the Automation Editor. So essentially all that's being done is calculating the relative pos of the mouse to the nearest note, and setting the drag value in the automation clip to that position and key.

You can still access the old automation editor version by shift-click (should I change it to be double-click?)

https://github.com/user-attachments/assets/50cc6ec0-a4af-4aa4-816d-786efb0e7ef9

(Apologies for the non-standard theme; I didn't feel like recording another demo)

![image](https://github.com/user-attachments/assets/09ed8ca3-89af-45f8-8dbf-9bbecf2661de)

@messmerd asked if the code could be made general for any future note parameter besides detuning, in preparation for CLAP per-note parameter automation. I have refactored the relevant functions to accept an enum type for the kind of note parameter, although currently only detuning is supported.

## Changes
- A new enum type was added, `Note::ParameterType`, which currently only has `Detuning`.
- New variables were added to handle the detuning in `PianoRoll.h`, to store the mouse states, the current notes being operated on, and the last changed tick.
- New functions were added to handle the detuning, for setting up the note vector, updating the drag position, and applying the drag position. Each of these functions take in a `Note::ParameterType` and use a switch statement to get the right automation clip in the note (currently only detuning is supported). Another function was also added to get the closest note to the mouse taking into account the shape of the automation curve.
- The drawing of the detuning info was changed to have circles around the automation nodes.
- The logic handling the detuning mode in the mouse event functions was updated.

regulus79 committed:
Initial Implementation

regulus79 committed:
Fix undo and bug with removing nodes

regulus79 committed:
Refactor to generalize to any future note parameter

regulus79 committed:
Add comment

regulus79 committed:
Fix formatting

sakertooth commented:
@szeli1 I believe you had similar ideas, but it was for manually editing automation clips directly in the Song Editor?

szeli1 commented:
> @szeli1 I believe you had similar ideas, but it was for manually editing automation clips directly in the Song Editor?

#7317
It is stuck because it doesn't improve the speed of automation. Shortcuts would fix this.

bratpeki commented:
What if I slide pitch from one note onto a place where another note already exists, and then click on that other note? If you don't get what I mean, I'll record a video demo when I can.

bratpeki commented:
Just tried it, it works on selections, so even if you overlap notes, there's no issue! _And_ it supports selections in the automation mode! Excellent! :grin:

bratpeki commented:
Mind the quality of the video, I overcompressed it. But I found a bug!


https://github.com/user-attachments/assets/7896719b-f996-4fc9-9b6a-02b2b56d060f

szeli1 commented on include/PianoRoll.h:
```diff
@@ -465,6 +468,14 @@ protected slots:
+	bool m_parameterEditDownRight = false;
+	int m_lastParameterEditTick = -1;
+
+	void updateParameterEditPos(QMouseEvent* me, Note::ParameterType paramType);
+	void applyParameterEditPos(QMouseEvent* me, Note::ParameterType paramType);
+	bool setupParameterEditNotes(Note::ParameterType paramType);
```
Please could you add some comments?

szeli1 commented on include/PianoRoll.h:
```diff
@@ -365,6 +365,7 @@ protected slots:
 	static const std::vector<float> m_zoomYLevels;
 
 	MidiClip* m_midiClip;
+	NoteVector m_selectedParameterEditNotes;
```
It would be better if `getSelectedNotes()` returned a reference to the selected notes. This way the code is more general and optimized. So essentially I'm asking you to implement something like this:

1. `NoteVector m_selectedNotes`
2. `void setSelectionChanged(Note*)` this will add or remove a note from `m_selectedNotes`, call this when a note is selected or unselected
3. `getSelectedNotes()` will return `m_selectedNotes`

I hope this will improve the code quality in the long run

regulus79 replied:
Are you asking me to change `PianoRoll::getSelectedNotes` or `m_selectedParameterEditNotes`? I have not touched `PianoRoll::getSelectedNotes`, so I'm not sure if that would be in the scope of this PR.

szeli1 replied:
I'm asking you to rename `m_selectedParameterEditNotes` to a generic name (I can't decide if it is already), make it always store all selected notes with a method called `void setSelectionChanged(Note*)`, and rewrite `getSelectedNotes()` so it returns `m_selectedParameterEditNotes`.

> I'm not sure if that would be in the scope of this PR.

I believe it will make future contributions easier, but it might be a bad idea to implement it here. Maybe you or I can refactor it in an other PR. The idea is to make it optimized (so it doesn't have to loop through all notes to return the selected ones) and make it (the selected notes) use a single vector.

regulus79 replied:
Oh you're right I realized `m_selectedParameterEditNotes` is always just the currently selected notes lol. So I could technically replace it with calls to `getSelectedNotes`. But I see why you want to refactor `getSelectedNotes`, it literally loops through all the notes to find the selected ones, which is probably not ideal. Caching them in a vector would definitely be more efficient (which is sort of what `m_selectedParameterEditNotes` does, but that's only for this feature).

bratpeki replied:
What's the consensus on this? Reviewing this right now, so I wanna know if this variable needs any changing.

szeli1 commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -2755,6 +2780,95 @@ void PianoRoll::mouseMoveEvent( QMouseEvent * me )
+	}
+	else if (noteUnderMouse())
+	{
+		m_selectedParameterEditNotes.assign(1, noteUnderMouse());
```
If you decide to refactor note selection, I believe you could just select manually that note here, so it is added to a variable called something like `m_selectedNotes`

regulus79 replied:
I should add a comment to clarify; the only purpose for selecting the note under the mouse is to keep track of it even if the user's cursor leaves the note while drawing the curve. It's a bit hacky but it works.

regulus79 replied:
But you are right, it is much simpler to just make the note selected and track it that way.

szeli1 commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -4718,6 +4832,54 @@ Note * PianoRoll::noteUnderMouse()
+	// Loop through all parameter edit notes
+	int minDifference = -1;
+	Note* closestNote = nullptr;
+	for(Note* note : m_selectedParameterEditNotes)
```
```suggestion
	for (Note* note : m_selectedParameterEditNotes)
```

szeli1 reviewed (changes requested):
I didn't test this PR, but I reviewed it.

regulus79 commented:
> But I found a bug!

Is the bug you are referring to the fact that after duplicating notes, changing one's detuning pattern changes the other? I believe that is also in master #5940 (but should be fixed in #7477)

regulus79 committed:
Simplify setting m_parameterEditDownRight

Co-authored-by: Sotonye Atemie <sakertooth@gmail.com>

regulus79 commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -2755,6 +2780,95 @@ void PianoRoll::mouseMoveEvent( QMouseEvent * me )
+		for (Note* note: m_selectedParameterEditNotes)
+		{
+			AutomationClip * aClip = nullptr;
+			switch (paramType)
+			{
+				case Note::ParameterType::Detuning:
+					aClip = note->detuning()->automationClip();
+					break;
+				default:
+					continue;
+			}
```
You're right, I should probably add a function which handles this instead of having the switch statement everywhere.

It is redundant, but @messmerd wanted me to make the code generalizable to other per-note parameter things in the future, aside from detuning.

sakertooth replied:
> It is redundant, but @messmerd wanted me to make the code generalizable to other per-note parameter things in the future, aside from detuning.

No need to future-proof if `Detuning` is the only parameter type. It just adds more work and resistance to the readability of the code.

regulus79 committed:
Simplify logic with dragging/removing nodes

Co-authored-by: Sotonye Atemie <sakertooth@gmail.com>

regulus79 committed:
Add semicolon

bratpeki commented:
@regulus79, could very well be! My exact steps were Ctrl+A, Shift-Dragging new notes into existence, and then attempting to change the pitch of those new notes, since they were selected afterwards. I'll try that on master and let you know if that's the same behavior.

regulus79 committed:
Add more comments

regulus79 committed:
Replace switch statement with single function call which handles returning the right automation clip (Currently only detuning is supported)

regulus79 commented:
I have added a lot of comments and refactored the switch statement out into a separate function. Hopefully the PR is easier to read now.

regulus79 commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -4718,6 +4832,54 @@ Note * PianoRoll::noteUnderMouse()
+		}
+	}
+	return closestNote;
+}
```
I have added comments and refactored out the switch statement, so hopefully it is better formatted now?

sakertooth replied:
Well that's refactoring, not formatting. I also don't like the verbosity of comments, especially ones that just repeat what the code is doing, but maybe that's just me.

sakertooth replied:
Example:

```cpp
// Get relative key
int relativeKey = key_num - note->key();
```
No value IMO

regulus79 replied:
I believe that excerpt is outdated; I have removed/added some comments and moved things around in a recent commit.

Rossmaxx commented:
Out of scope for this PR, i see a chance we can replace `lmms::shared_object` to `std::shared_ptr`. Afaik shared_object is used only for detuning. I was supposed to do it for #7677 but was fearing a change in functionality and a few build fails. I am not in a position to test any changes.

bratpeki commented:
@regulus79, the named issue definitely exists on master. Adding it to my notes and opening a corresponding issue if there isn't one.

regulus79 committed:
Fix formatting

regulus79 committed:
Draw circles on top of lines

regulus79 committed:
Change switch intentation and comment

regulus79 committed:
Fix middle-mouse-button issues

regulus79 committed:
Change m_parameterEditDown to m_parameterEditDownLeft and add some checks to prevent holding left and right at the same time

bratpeki commented:
I see the MMB issue has been fixed. Saving and loading works. Please merge ASAP! :sunglasses:

regulus79 commented:
I just realized that currently the detuning is locked to whole semitone steps. I guess that's fine, but if velocity/panning/whatever per-note parameter automation is implemented, that might have to change.

sakertooth commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -4718,6 +4845,50 @@ Note * PianoRoll::noteUnderMouse()
+
+Note * PianoRoll::parameterEditNoteUnderMouse(Note::ParameterType paramType)
+{
+	QPoint pos = mapFromGlobal( QCursor::pos() );
```
```suggestion
	QPoint pos = mapFromGlobal(QCursor::pos());
```

sakertooth commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -4718,6 +4845,50 @@ Note * PianoRoll::noteUnderMouse()
+{
+	QPoint pos = mapFromGlobal( QCursor::pos() );
+
+	if (pos.x() <= m_whiteKeyWidth
+		|| pos.x() > width() - SCROLLBAR_SIZE
+		|| pos.y() < PR_TOP_MARGIN
+		|| pos.y() > keyAreaBottom() )
+	{
+		return nullptr;
+	}
```
```suggestion
	if (pos.x() <= m_whiteKeyWidth
		|| pos.x() > width() - SCROLLBAR_SIZE
		|| pos.y() < PR_TOP_MARGIN
		|| pos.y() > keyAreaBottom())
	{
		return nullptr;
	}
```

sakertooth commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -4718,6 +4845,50 @@ Note * PianoRoll::noteUnderMouse()
+	}
+
+	// Get the key and time pos of the cursor
+	int key_num = getKey( pos.y() );
```
```suggestion
	int key_num = getKey(pos.y());
```

sakertooth commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -4718,6 +4845,50 @@ Note * PianoRoll::noteUnderMouse()
+
+	// Get the key and time pos of the cursor
+	int key_num = getKey( pos.y() );
+	int pos_ticks = (pos.x() - m_whiteKeyWidth) *
```
We use camelCase, not snake case.

regulus79 replied:
I think I copied this code from PianoRoll::noteUnderMouse, which used that format. I can change it to camelCase though.

regulus79 committed:
Fix style

Co-authored-by: Sotonye Atemie <satemiej@gmail.com>

sakertooth commented:
Cool feature, thanks. The feature only allows detuning up until the max length of the note, which I think wasn't the same as when it was done through the editor. I also think its a problem because you can have release on a note, as to which the user may want to apply the detuning beyond the length of the note.

sakertooth commented:
Also wanted to mention something: if you feel the need to add comments (excluding documentation comments) everywhere, it's always a good idea to see if you can improve your naming and simplify your interfaces. It will help developers understand what's going on better. Try not to use comments as a crutch.

regulus79 committed:
Allow pitchbending past end of note

regulus79 committed:
Fix visual bug and change automation node radius

regulus79 committed:
Show textfloat at start to give instructions

regulus79 commented:
> The feature only allows detuning up until the max length of the note, which I think wasn't the same as when it was done through the editor. I also think its a problem because you can have release on a note, as to which the user may want to apply the detuning beyond the length of the note.

Okay, so...... This is a bit more difficult to implement, since previously I was relying on the vertical distance form the mouse to the detuning curve to determine the closest note, but this would require also calculating the distance to the last node in the curve.

But, after a while, I actually did get it to work fairly well https://github.com/LMMS/lmms/pull/7759/commits/4dbcf93f782b3b394b5b038c69c511fd9d1d08b9. So now you can pitchbend past the end of the note!


On a separate note, I also changed the color and size of the automation nodes, and I added a `TextFloat` which appears for 4 seconds when you enter detuning mode to give instructions, in particular how you can shift-click to view the automation editor version.

![image](https://github.com/user-attachments/assets/e2b3bc0f-f6d4-48ab-b035-4cda356b2935)

regulus79 committed:
Merge branch 'master' into pianoroll-pitchbending

bratpeki commented:
I think the addition of an infobox is great, but I think the current text could cause confusion.

---

_Click on a note [...] to edit its detuning curves [...]_ could incentivize that clicking begins the editing process, whereas that's "not true" and it actually places a point immediately.

So say someone reads that, clicks on a note to edit it, and immediately places a point they didn't want, so they have to undo.

---

Of course, the user has to get adjusted to the fact that they actually click-and-drag the point into place, but maybe there's a better way to word this?

bratpeki commented:
Yeah, the more I think about it, the more I think we just have to incentivize the "click-and-drag" idea, the rest of the text is good as-is!

regulus79 commented:
> This may not be the right place to discuss this, but I think the bending automation editor is somewhat bugged, since it places the notes off-center, for some reason.

I kind of made a PR to fix that a while ago, but I closed it in favor of this one #7465

bratpeki commented:
If we are keeping the bend automation editor, and a fix exist, I'm all for merging it upstream!

rdrpenguin04 commented:
Keeping the automation editor could help for pitch bends with non-slope shapes, so unless this editor also implements those, I think the automation editor is valuable enough to keep around. At least for now.

bratpeki commented:
Actually, great point, @rdrpenguin04! Thank you, that slipped my mind! :pray: 

I noticed we don't display those curves, it's probably a performance thing. @regulus79 might have more info.

bratpeki commented:
Also, @regulus79, have you given a different text in the infobox any thought? Are you for keeping this one or focusing on the "click-and-drag" idea more?

regulus79 commented:
> Keeping the automation editor could help for pitch bends with non-slope shapes, so unless this editor also implements those, I think the automation editor is valuable enough to keep around. At least for now.

Yes, that was kind of what I was thinking. Since changing between discrete, linear, and cubic curves, and changing tangents, or out/in values are not implemented in this PR, we need to have the automation editor version still accessible for advanced editing.

> I noticed we don't display those curves, it's probably a performance thing. regulus79 might have more info.

Discrete and linear curves should display just fine in the piano roll, but as of now no one has yet implemented proper drawing for cubic pitch bending curves. Aside from adding the circles, I did not change the drawing code in this PR.

> Also, regulus79, have you given a different text in the infobox any thought? Are you for keeping this one or focusing on the "click-and-drag" idea more?

I could change it to something like "Drag on a note or selection to edit their detuning curves", or maybe you have ideas on what it should say?

regulus79 committed:
Update textfloat

bratpeki commented:
Nice! I'd just like to notice it's _one curve,_ but _multiple points_. How about:

"Click and drag on a note/selection/point to edit the detuning curve"

?

regulus79 committed:
Change 'curves' to 'curve'

regulus79 commented:
You're right, "curves" should be "curve" since it's just talking about a single note, thanks for catching that.

szeli1 reviewed (changes requested):
I have noticed some issues:
1. I can drag the automation upwards, then I can move the note to the absolute top and the pitch moves out of the piano roll. I don't know if this is intended. You could ignore this if this is intended.
2. If I move the automation node / point left 2-3 bars, if the note moves off screen, the automation node disappears. I think it would be better if the note / chord would grow left as the node is dragged out of the chord.
3. Select a chord, place 3 automation nodes beside each other. Click on the middle one and drag up or downwards. Once the top / bottom node hits the editor's top / bottom side, the nodes jump downward / upward (while still dragging). You could ignore this if its hard to fix.

I think this PR is a huge improvement compared to the automation editor method. I would like to get this merged.

regulus79 commented:
Tysm for testing! :D

> I can drag the automation upwards, then I can move the note to the absolute top and the pitch moves out of the piano roll. I don't know if this is intended. You could ignore this if this is intended.

This was not intended, but I'm not sure how to fix it. At worst, if the user needs to edit those nodes, they can always open the note in the normal automation editor and adjust it there. Then again, pitching bending a note outside of the piano bounds seems like something which shouldn't be supported? But whatever, maybe it's fine to leave it in.

> If I move the automation node / point left 2-3 bars, if the note moves off screen, the automation node disappears. I think it would be better if the note / chord would grow left as the node is dragged out of the chord.

Moving the automation nodes 2-3 bars left? Did you mean right? Afaik you shouldn't be able to move the automation nodes before the start of the note. If you did mean moving the nodes past the right end of the note, originally I had it so that they weren't able to do that, but saker (I think) pointed out that this means you wouldn't be able to pitch bend the release of a note, since the automation curve would stop at the end of the note. So I made it so that you can pitch bend after the note ends.

> Select a chord, place 3 automation nodes beside each other. Click on the middle one and drag up or downwards. Once the top / bottom node hits the editor's top / bottom side, the nodes jump downward / upward (while still dragging). You could ignore this if its hard to fix.

Hmm, I can't seem to reproduce this. I placed three nodes one after each other, then dragged the middle one to the top/bottom of the piano roll, but nothing interesting happened. (Well, actually when you drag the node down and your mouse runs into the velocity part, you start changing the velocity.... but that bug exists in master with the knife tool too sooo)

regulus79 committed:
Fix style

Co-authored-by: szeli1 <143485814+szeli1@users.noreply.github.com>

szeli1 commented:
> Moving the automation nodes 2-3 bars left? Did you mean right? Afaik you shouldn't be able to move the automation nodes before the start of the note. If you did mean moving the nodes past the right end of the note, originally I had it so that they weren't able to do that, but saker (I think) pointed out that this means you wouldn't be able to pitch bend the release of a note, since the automation curve would stop at the end of the note. So I made it so that you can pitch bend after the note ends.

Sorry I meant right. If you drag the node right really far, then use the horizontal scroll bar to scroll right, once the note goes off screen, the node disappears.

> Hmm, I can't seem to reproduce this. I placed three nodes one after each other, then dragged the middle one to the top/bottom of the piano roll, but nothing interesting happened.

Place 3 notes above each other and select them. Use the automation tool to make 3 pitch bending nodes for each note. Now you should have 9 nodes. Start dragging the one at the center up and down. If you drag it upwards and the top node hits the window's top side, then the nodes should jump downwards. I think this behavior feels like a glitch

regulus79 committed:
Fix detuning not ebing drawn when note goes out of view

regulus79 committed:
Store the current note so that it doesn't need to be recalculated every time, to prevent the odd switching between curves when dragging.

regulus79 commented:
> Sorry I meant right. If you drag the node right really far, then use the horizontal scroll bar to scroll right, once the note goes off screen, the node disappears.

Oh, now I see. That is definitely a problem. I think it's due to notes (and their detuning curves) only being drawn if they are visible. This is also a bug in master. But I think I've fixed it now!

> Place 3 notes above each other and select them. Use the automation tool to make 3 pitch bending nodes for each note. Now you should have 9 nodes. Start dragging the one at the center up and down. If you drag it upwards and the top node hits the window's top side, then the nodes should jump downwards. I think this behavior feels like a glitch

Now I see what you are talking about. I agree it does feel a bit odd, though I think it's mainly due to how I am constantly calculating which detuning curve is vertically closest to the mouse, so when you have notes very close to each other, it can sometimes flip between the curves.

I just got an idea though; if I only calculate the closest curve when the user first presses the mouse to start the drag, then that should prevent the "focus" from switching between curves.

WOW now that I've implemented that in the most recent commit, it feels awesome. So much smoother! Thanks for reporting that bug!

regulus79 committed:
Merge branch 'master' into pianoroll-pitchbending

regulus79 committed:
Update formatting

regulus79 commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -3627,11 +3761,14 @@ void PianoRoll::paintEvent(QPaintEvent * pe )
 
 			int pos_ticks = note->pos();
 
+			int detuning_length = note->detuning()->automationClip()->getTimeMap().lastKey() * m_ppb / TimePos::ticksPerBar();
```
I have the formatting

regulus79 committed:
Merge branch 'master' into pianoroll-pitchbending

regulus79 committed:
Fix crash when detuning curve is empty

szeli1 reviewed (commented):
I will test this PR soon.

regulus79 committed:
Fix style and move comment to two lines

Co-authored-by: szeli1 <143485814+szeli1@users.noreply.github.com>

regulus79 committed:
Merge branch 'master' into pianoroll-pitchbending

regulus79 committed:
Fix indent

regulus79 committed:
Fix new nodes being added when clicking on unselected note which already has detuning

regulus79 commented:
Thanks for testing!

> Either a new node will be created where the user clicked or even worse an existing node will be moved. I would expect to enter into the pitch bending mode without modifying the nodes.

Thanks for catching this. It should be fixed now.

> Automation nodes go out of bound.

Yeah, I see that. Unfortunately it's not an easy thing to fix. However, the user can always use shift+click to open the note in the automation editor, so they can still edit the detuning.

> Once the user scrolls to the top, clicking on nodes becomes weird. It's like the cursor clicks things that are above it. If the user meant to click on a note, a new note is created instead (above the visually clicked note). This is not present in nightly.

This is very weird. I can reproduce it both on this PR and on an appimage build of a different PR. It only occurs after you click to change the mode from the top toolbar. It seems like when you interact with those buttons for the first time, they grow slightly (or maybe not idk), which must be messing up the pixel calculations for the notes.

Before you click on a button:
![image](https://github.com/user-attachments/assets/1be58f57-5251-4802-820d-eb8700a9ba64)

After:
![image](https://github.com/user-attachments/assets/24c4dc15-cd25-4877-bbd9-38a7d9b7c25a)

(Those screenshots aren't very useful actually, ignore them lol.) I didn't quite catch it in that screenshot, but for a split second the piano shifted a bit and then went back. Now all of a sudden you have the ability to scroll up one more step. If you scroll up, nothing changes visually, but now it thinks your mouse is at a different offset. So after interacting with those toolbar buttons, the piano roll gains an extra vertical scroll step at the top, which causes problems it seems(?)

Interestingly if you close the piano roll and open it again, the issue goes away and you can't get it to re-occur. So the piano roll somehow gets reset... but not. This is odd.

But in any case, It doesn't seem this bug is exclusive to this PR. And I didn't touch any code relating to normal selection of notes, so I don't believe I caused it.

bratpeki commented:
Gonna daily drive this today and if it's all good, final checks and merge!

sakertooth reviewed (commented):
The feature works well for the most part, but doesn't draw the curves correctly when its a non-linear one.

sakertooth commented:
> I believe this is an issue in master as well.

I mean the new feature doesn't draw the curves the same way it is drawn in the editor.

bratpeki commented:
<img width="720" height="540" alt="2025-07-28-184425_720x540_scrot" src="https://github.com/user-attachments/assets/95677234-bc8c-4ae1-9c25-b11e15fe2213" />

@sakertooth; can confirm.

bratpeki commented:
<img width="878" height="491" alt="2025-07-28-184631_878x491_scrot" src="https://github.com/user-attachments/assets/af37d5f9-1773-4908-8525-fa288a583b5c" />

@regulus79; can also confirm the curve is not properly drawn on master.

bratpeki commented:
I think that's a fix for a future PR then, so I say this is still tested.

regulus79 committed:
Merge branch 'master' into pianoroll-pitchbending

regulus79 committed:
Remove get prefix and add/change some comments

regulus79 committed:
Update TextFloat message

regulus79 commented:
Is this PR ready to merge? I don't see any important open code reviews, and I can't recall any outstanding issues which need to be fixed.

regulus79 commented on include/Note.h:
```diff
@@ -123,6 +124,12 @@ class LMMS_EXPORT Note : public SerializingObject
+	{
+		Detuning = 0
+	};
+	AutomationClip* getParameterCurve(ParameterType paramType);
```
Oh whoops, I missed this review comment, sorry. I will try to remove all "get" prefixes.

bratpeki commented on include/Note.h:
```diff
@@ -123,6 +124,13 @@ class LMMS_EXPORT Note : public SerializingObject
 	Type type() const { return m_type; }
 	inline void setType(Type t) { m_type = t; }
 
+	//! Types of per-note automation. Currently only detuning/pitch bending is supported.
+	enum class ParameterType
+	{
+		Detuning = 0
+	};
+	AutomationClip* parameterCurve(ParameterType paramType);
```
Could you elaborate a bit on this? Why forward declare `AutomationClip` in `Note`?

I'm also looking at [`include/AutomationClip.h` (in your branch)](https://github.com/regulus79/lmms/blob/pianoroll-pitchbending/include/AutomationClip.h), it doesn't seem to define any CTOR that uses `ParameterType`. What's that about?

regulus79 committed:
Merge branch 'master' into pianoroll-pitchbending

regulus79 committed:
Fix style

regulus79 committed:
Update comments

regulus79 committed:
Don't let the user accidentally delete first node

regulus79 committed:
Tweak comment wording

regulus79 committed:
Fix issue with dragging first node

regulus79 committed:
Merge branch 'master' into pianoroll-pitchbending

regulus79 committed:
Fix bug with holding ctrl while dragging, then releasing mouse, then releasing ctrl

regulus79 committed:
Remove m_selectedParameterEditNotes and use getSelectedNotes() instead

sakertooth reviewed (approved):
Could've made a couple of nitpicks but don't want to withhold this much needed feature. Tested it and it works mostly fine, but found a minor bug (?) where the green line disconnects once moved to the edge near the keys in the Piano Roll. Not sure if you want to look into that here or later (or at all).

regulus79 commented:
Thanks for approving!

> found a minor bug (?) where the green line disconnects once moved to the edge near the keys in the Piano Roll

I'm not sure I understand what bug you are referring to. I did notice a similar potential bug where dragging a node past the left edge of the note and dragging it back without releasing the mouse could cause it to update the height of the first node, even though it wasn't explicitly clicked and dragged. That's mainly due to the hacky way I'm using the automation clip editing functions to disallow the user from dragging the first node left/right. I think it's a pretty minor bug (and would probably be tricky to solve without a lot of reworking) so it might be okay to leave it for now. Though I'm not sure if that's the same bug you are referring to.

sakertooth commented:
> Thanks for approving!
> 
> > found a minor bug (?) where the green line disconnects once moved to the edge near the keys in the Piano Roll
> 
> I'm not sure what bug you are referring to. I did notice a similar potential bug where dragging a node past the left edge of the note and dragging it back without releasing the mouse could cause it to update the height of the first node, even though it wasn't explicitly clicked and dragged. That's mainly due to the hacky way I'm using the automation clip editing functions to disallow the user from dragging the first node left/right. I think it's a pretty minor bug (and would probably be tricky to solve without a lot of reworking) so it might be okay to leave it for now. Though I'm not sure if that's the same bug you are referring to.


https://github.com/user-attachments/assets/e9ff39e7-c85a-4670-8979-83d29d41577a

I think this is the same bug you're mentioning, but just for clarify I attached a video. I think we should fix this before merge?

regulus79 committed:
Fix issues with dragging around first node

regulus79 commented:
I've fixed the bug. I also made it smoother to use, so instead of the old drag ending and a new drag starting, it just kind of clamps the drag but still lets you move it up and down. It's hard to explain, but I think it feels better.


https://github.com/user-attachments/assets/992318f8-e8d6-4107-8574-80b4edec9dc7

regulus79 committed:
Remove whitespace

regulus79 committed:
Use std::optional for m_lastParameterEditTick

sakertooth commented:
> I've fixed the bug. I also made it smoother to use, so instead of the old drag ending and a new drag starting, it just kind of clamps the drag but still lets you move it up and down. It's hard to explain, but I think it feels better.
> 2026-01-04.10-35-58.mp4

Great, thats a bit better, but I think the first node should always be attached to the note right?

regulus79 committed:
Fix bug with clicking unselected notes

regulus79 commented:
Are we good to merge?

bratpeki commented:
IMO, yes.

regulus79 commented:
> Will this new tool support 0.5 semitone bending in the future, like the automation editor (or even finer tuning, for that matter)?

It probably wouldn't be extremely hard to add, perhaps letting the user hold down alt or ctrl or something. For vertical un-quantized movement, that's fine. But for the horizontal un-quantized adjustment, the current implementation where it essentially interfaces with the functions used by the automation editor means that it's a bit more tricky to get that to work.

SCPlayz7000 commented:
This is like watching the mona lisa getting painted

SCPlayz7000 commented:
you guys don't know how much this means to me ❤️‍🔥️
